etl/steps/__init__.py

```diff
@@ -656,7 +668,11 @@ def run(self) -> None:
 
         with _unignore_backports(Path(self._path)):
             try:
-                self._cached_dvc_repo().pull(self._path, remote="public-read", force=True)
+                repo = self._cached_dvc_repo()
+                # DVC must be locked across processes. This is pretty limiting as it allows us to only pull
+                # one snapshot at a time. One option is to pre-pull all snapshot steps.
+                with dvc_pull_lock:
+                    repo.pull(self._path, remote="public-read", force=True)
```

Comment:
This turns out to be a big slow-down if you're building the ETL from scratch. If we wanted to speed this up as well, we'd need to preferentially execute non-snapshot steps whenever they're available.

This is more of a note for the future. See earlier comment about the ordering of steps into the topological sorter.

Reply:
True. I'm honestly a bit disappointed by DVC and its multiprocessing. In the future we might consider writing custom pull function (adapted from walden) or replace the entire DVC.

(In nightly builds when building ETL from scratch we `dvc pull` all files before running the actual ETL. That happens in parallel and is pretty fast, like 5 mins.)